table/block_based/partitioned_index_reader.cc

```diff
@@ -173,7 +173,7 @@ Status PartitionIndexReader::CacheDependencies(const ReadOptions& ro,
       return s;
     }
     if (block.GetValue() != nullptr) {
-      if (block.IsCached()) {
+      if (block.IsCached() || block.GetOwnValue()) {
```

Comment:
This fixes another edge case leading to corruption. Due to how `partition_map_` is used, it must contain either all partitions or no partitions. When it contains only some partitions, the partitions not in the map are assumed to be empty (https://github.com/facebook/rocksdb/blob/12b78e40bdba980f3692fd2018cc905eef2127f8/table/block_based/block_based_table_reader.cc#L1977-L1993). Such a wrong assumption causes corrupt reads.

One case in which only some partitions would be added to the map is when `enable_index_compression == 1`. It can be the case that some of the partitions are compressed while others are not (e.g., due to ineffective compression ratio). The compressed partitions would be decompressed, loaded to block cache, and then inserted into `partition_map_`. Meanwhile, the uncompressed partitions could be read directly from the file, e.g., when `mmap_read == true`, and such blocks (until this PR) did not get inserted into `partition_map_`.